### ISSUE TITLE
Training an RNN to generate a sine wave

### ISSUE THREAD

**Issue by DarylWM** [tflearn/tflearn#121]
I'm attempting to do this but the predicted values (in green) don't look anything like a sine wave.

![image](https://cloud.githubusercontent.com/assets/1016303/15698647/fac9e1ae-2809-11e6-8904-f96fa0a27d68.png)

I've tried various learning rates, optimizers, number of units, length of history for training, number of steps I try to look ahead. I'm not yet seeing what the problem is, so I would be grateful for suggestions.

```
# Simple example using recurrent neural network to predict time series values

from __future__ import division, print_function, absolute_import

import tflearn
from tflearn.layers.normalization import batch_normalization
import numpy as np
import math
import matplotlib
matplotlib.use('Agg')
import matplotlib.pyplot as plt

step_radians = 0.01
steps_of_history = 100
steps_in_future = 2
index = 0

x = np.arange(0, 4*math.pi, step_radians)
y = np.sin(x)

# Put the data into the right shape
while (index+steps_of_history+steps_in_future < len(y)):
	window = y[index:index+steps_of_history]
	target = y[index+steps_of_history+steps_in_future]
	if index == 0:
		trainX = window
		trainY = target
	else:
		trainX = np.vstack([trainX, window])
		trainY = np.append(trainY, target)
	index = index+1
trainX.shape = (index, steps_of_history, 1)
trainY.shape = (index, 1)

# Network building
net = tflearn.input_data(shape=[None, steps_of_history, 1])
net = tflearn.simple_rnn(net, n_units=512, return_seq=False)
net = tflearn.dropout(net, 0.5)
net = tflearn.fully_connected(net, 1, activation='linear')
net = tflearn.regression(net, optimizer='sgd', loss='mean_square', learning_rate=0.001)

# Training
model = tflearn.DNN(net, clip_gradients=0.0, tensorboard_verbose=0)
model.fit(trainX, trainY, n_epoch=150, validation_set=0.1, show_metric=True, batch_size=128)

# Prepare the testing data set
# testX = window to use for prediction
# testY = actual value
# predictY = predicted value
index = 0
while (index+steps_of_history+steps_in_future < len(y)):
	window = y[index:index+steps_of_history]
	target = y[index+steps_of_history+steps_in_future]
	if index == 0:
		testX = window
		testY = target
	else:
		testX = np.vstack([testX, window])
		testY = np.append(testY, target)
	index = index+1
testX.shape = (index, steps_of_history, 1)
testY.shape = (index, 1)

# Predict the future values
predictY = model.predict(testX)

# Plot the results
plt.figure(figsize=(20,4))
plt.suptitle('Prediction')
plt.title('History='+str(steps_of_history)+', Future='+str(steps_in_future))
plt.plot(y, 'r-', label='Actual')
plt.plot(predictY, 'gx', label='Predicted')
plt.legend()
plt.savefig('sine.png')

```

**Comment by aymericdamien** [comment]
Yes, it seems it is always predicting same value. It might come from your training data. When you are training your model, does the loss is going down (which means your model is actually learning something)?

Also it might be better to use LSTM for such task, and you can try to stack some rnn layers:
```
net = tflearn.input_data(shape=[None, steps_of_history, 1])
net = tflearn.lstm(net, n_units=128, return_seq=True)
net = tflearn.lstm(net, n_units=128, return_seq=True)
net = tflearn.lstm(net, n_units=128, return_seq=False)
net = tflearn.fully_connected(net, 1, activation='linear')
net = tflearn.regression(net, optimizer='sgd', loss='mean_square', learning_rate=0.001)
```

**Comment by DarylWM** [comment]
Thanks - I tried your suggested LSTM-based model. I've tried LSTM on this problem before but didn't have three layers. The validation loss increases during training, so clearly I'm doing something wrong with the training data.

![image](https://cloud.githubusercontent.com/assets/1016303/15728813/520af06e-28a2-11e6-9c9e-bf714b0f4503.png)

To make sure I'm feeding the training data to the model properly, I have the sine values in the array `y`. The training set is a series of 100 consecutive values from `y`, with the target being the 102nd value.

For example, the training pairs should be:
X1 = y[0:100]
Y1 = y[102]

X2 = y[1:101]
Y2 = y[103]

and so on.

So for `model.fit()`:
```
trainX[number_of_training_pairs][100][1]    # sequence of 100 consecutive sine values
trainY[number_of_training_pairs]            # target value, two sine values further
```

Is that the right way to feed it?

**Comment by aymericdamien** [comment]
The issue is maybe related to your data.
I am not so clear what you are trying to do. It seems you want to input a sequence of points, and predict the rest of the function, so that if you input sin([0: 10]), the network should be able to predict sin([10:20]) ? is that right?

**Comment by DarylWM** [comment]
I'm using a sine wave as a learning exercise for applying an RNN to time series prediction. I want to input sin([0:10]) and get the model to predict sin([12]). Just a single value in the 'future'.

**Comment by aymericdamien** [comment]
Hi, I updated  your example, I am not sure if it was a problem. Also your data range was quite small, a larger dataset helped, also in such example we could set a higher learning rate.

```
# Simple example using recurrent neural network to predict time series values

from __future__ import division, print_function, absolute_import

import tflearn
from tflearn.layers.normalization import batch_normalization
import numpy as np
import math
import matplotlib
matplotlib.use('Agg')
import matplotlib.pyplot as plt

step_radians = 0.01
steps_of_history = 200
steps_in_future = 1
index = 0

x = np.sin(np.arange(0, 20*math.pi, step_radians))

seq = []
next_val = []

for i in range(0, len(x) - steps_of_history, steps_in_future):
    seq.append(x[i: i + steps_of_history])
    next_val.append(x[i + steps_of_history])

seq = np.reshape(seq, [-1, steps_of_history, 1])
next_val = np.reshape(next_val, [-1, 1])
print(np.shape(seq))

trainX = np.array(seq)
trainY = np.array(next_val)

# Network building
net = tflearn.input_data(shape=[None, steps_of_history, 1])
net = tflearn.simple_rnn(net, n_units=32, return_seq=False)
net = tflearn.fully_connected(net, 1, activation='linear')
net = tflearn.regression(net, optimizer='sgd', loss='mean_square', learning_rate=0.1)

# Training
model = tflearn.DNN(net, clip_gradients=0.0, tensorboard_verbose=0)
model.fit(trainX, trainY, n_epoch=15, validation_set=0.1, batch_size=128)

# Testing
x = np.sin(np.arange(20*math.pi, 24*math.pi, step_radians))

seq = []

for i in range(0, len(x) - steps_of_history, steps_in_future):
    seq.append(x[i: i + steps_of_history])

seq = np.reshape(seq, [-1, steps_of_history, 1])
testX = np.array(seq)

# Predict the future values
predictY = model.predict(testX)
print(predictY)

# Plot the results
plt.figure(figsize=(20,4))
plt.suptitle('Prediction')
plt.title('History='+str(steps_of_history)+', Future='+str(steps_in_future))
plt.plot(x, 'r-', label='Actual')
plt.plot(predictY, 'gx', label='Predicted')
plt.legend()
plt.savefig('sine.png')
```

![sine](https://cloud.githubusercontent.com/assets/10386605/15736504/ae0a83d6-28d2-11e6-94f9-29e0ea37703b.png)

**Comment by DarylWM** [comment]
Thanks @aymericdamien. You are correct - increasing the data set, from 4*pi to 20*pi, and increasing the number of historical values used for each training pair, from 100 to 200, made it work.

One thing about your code - I don't think it's training to predict a value in the future; trainY seems to contain the last value of the corresponding trainX sequence. Predicting ahead a few steps does work though.

**Comment by aymericdamien** [comment]
TrainY is actually the next value, because:
```
# if x=[0, 1, 2, 3], steps_of_history = 4, i = 0
seq.append(x[i: i + steps_of_history]) # x[0:4] = [0, 1, 2]
next_val.append(x[i + steps_of_history]) # x[4] = [3]
```
so in fact it should be:
```
next_val.append(x[i + steps_of_history + steps_in_future - 1])
```

Also I think that even setting the steps_of_history to 10 can still make it works, what was important was more training data.

**Comment by DarylWM** [comment]
Yes, sorry, I misread the range.

Even training on a history of 5 will get surprisingly close results.

![image](https://cloud.githubusercontent.com/assets/1016303/15766159/0bd17982-297f-11e6-9841-335e39324bd9.png)

Thanks for your help.

**Comment by aymericdamien** [comment]
Yes, RNN is working pretty well!

**Comment by crispy-penguin** [comment]
Hi everyone, 
I was trying to replicate your results but I was unable to do so.
The rnn is producing good results until epoch 3.
After 3-4 epochs the total loss just skyrockets to oblivion. 

`Training Step: 43  | total loss: 0.48619
| SGD | epoch: 001 | loss: 0.48619 | val_loss: 0.54735 -- iter: 5475/5475
--
Training Step: 86  | total loss: 0.45128
| SGD | epoch: 002 | loss: 0.45128 | val_loss: 0.36521 -- iter: 5475/5475
--
Training Step: 129  | total loss: 0.17766
| SGD | epoch: 003 | loss: 0.17766 | val_loss: 0.10661 -- iter: 5475/5475
--
Training Step: 172  | total loss: nan39457634632105965698262809706496.00000
| SGD | epoch: 004 | loss: nan | val_loss: inf -- iter: 5475/5475
--
Training Step: 215  | total loss: nan
| SGD | epoch: 005 | loss: nan | val_loss: nan -- iter: 5475/5475
`
Any ideas?
Cheers

**Comment by crispy-penguin** [comment]
For anyone with the problem above. Switching from cpu computation to gpu computation fixed the problem. Somehow.

**Comment by amityo9** [comment]
I tweaked learning rate to reduce the loss. I will post full code in a while. Thanks!

**Comment by amityo9** [comment]
# Simple example using recurrent neural network to predict time series values

from __future__ import division, print_function, absolute_import

import tflearn
import numpy as np
import math
import matplotlib
matplotlib.use('Agg')
import matplotlib.pyplot as plt
import tensorflow as tf

step_radians = 0.001
steps_of_history = 10
steps_in_future = 5
learning_rate = 0.003

def getData(x):
    seq = []
    next_val = []
    for i in range(0, len(x) - steps_of_history - steps_in_future, steps_in_future):
        seq.append(x[i: i + steps_of_history])
        next_val.append(x[i + steps_of_history + steps_in_future -1])
    
    seq = np.reshape(seq, [-1, steps_of_history, 1])
    next_val = np.reshape(next_val, [-1, 1])
    X = np.array(seq)
    Y = np.array(next_val)
    return X,Y

def myRNN(activator,optimizer):
    tf.reset_default_graph()
    # Network building
    net = tflearn.input_data(shape=[None, steps_of_history, 1])
    net = tflearn.lstm(net, 32, dropout=0.8,bias=True)
    net = tflearn.fully_connected(net, 1, activation=activator)
    net = tflearn.regression(net, optimizer=optimizer, loss='mean_square', learning_rate=learning_rate)
    
    # Training Data
    trainVal = np.sin(np.arange(0, 20*math.pi, step_radians))
    trainX,trainY = getData(trainVal)
    print(np.shape(trainX))
    
    # Training
    model = tflearn.DNN(net)
    model.fit(trainX, trainY, n_epoch=10, validation_set=0.1, batch_size=128)
    
    # Testing Data
    testVal = np.sin(np.arange(20*math.pi, 24*math.pi, step_radians))
    testX,testY = getData(testVal)
    
    # Predict the future values
    predictY = model.predict(testX)
    
    print("---------TEST ERROR-----------")
    expected = np.array(testY).flatten()
    predicted = np.array(predictY).flatten()
    error = sum(((expected - predicted) **2)/len(expected))
    print(error)
    
    # Plot and save figure
    plotFig(testY, np.array(predictY).flatten(), error, activator+"_"+optimizer)

def plotFig(actual,predicted,error,filename):
    # Plot the results
    plt.figure(figsize=(20,4))
    plt.suptitle('Prediction')
    plt.title('History = '+str(steps_of_history)+', Future = '+str(steps_in_future)+', Error= '+str(error*100)+'%')
    plt.plot(actual, 'r-', label='Expected')
    plt.plot(predicted, 'g.', label='Predicted')
    plt.legend()
    plt.savefig(filename+'.png')
    
def main():
    activators = ['linear', 'tanh', 'sigmoid', 'softmax', 'softplus', 'softsign', 'relu', 'relu6', 'leaky_relu', 'prelu', 'elu']
    optimizers = ['sgd', 'rmsprop', 'adam', 'momentum', 'adagrad', 'ftrl', 'adadelta']
    for activator in activators:    
        for optimizer in optimizers:
            print ("Running for : "+ activator + " & " + optimizer)
            myRNN(activator, optimizer)

main()

**Comment by imswiming** [comment]
can someone please explain the getData(x) function?

specifically this part:

```
for i in range(0, len(x) - steps_of_history - steps_in_future, steps_in_future):
        seq.append(x[i: i + steps_of_history])
        next_val.append(x[i + steps_of_history + steps_in_future -1])
    
    seq = np.reshape(seq, [-1, steps_of_history, 1])
    next_val = np.reshape(next_val, [-1, 1])
```

**Comment by baonlq** [comment]
Thank you! This is helpful for me.
By the way, getData func format the inputs and labels for input. each i in list 0, 5, 10, ... seq will be appended an array has length: steps_of_history - data returned from x (x is np.sin function ).
